### PR TITLE
Implement tower attack action 

### LINKED ISSUE #24: Implement tower attack action
TowerController has an attack function, which has to spawn a bullet. The bullet should be its own class (possibly full MVC pattern) and travel towards its target. When the target is hit, the a damage method must on UnitController must be called and the bullet controller (which extends Updatable) must call its die method. Note: The bullet must be a guaranteed hit.

### PR DESCRIPTION
**Pull request by fredrikbw**
Fix #24

### CODE REVIEW

File: client/core/src/com/cows/game/controllers/BulletController.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    override fun update(deltaTime: Float) {
+        if (model.hasHit) return
+        if (currentTarget == null) return
+        val targetPos = currentTarget!!.model.position
+        val oldX = roundNum(model.position.x)
+        val oldY = roundNum(model.position.y)
+
+        model.position.x += ((targetPos.x + 32f - model.position.x)) / model.movementSpeed
+        model.position.y += ((targetPos.y + 32f - model.position.y))/ model.movementSpeed
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/controllers/BulletController.kt:21]: Consistent spacing :(

Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/controllers/BulletController.kt:21]: Consistent spacing :(

File: client/core/src/com/cows/game/controllers/BulletController.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            BulletView(model)
+        }
+    }
+
+    override fun die() {
+        super.die()
+        model.hasHit = true
+    }
+
+    private fun roundNum(f: Float): Int {
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/controllers/BulletController.kt:40]: Is this for comparing floats? We could do a reachedTarget method which checks if delta between two floats is lower than an epsilon value? That's a more traditional way of comparing floats.

Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/controllers/BulletController.kt:40]: Is this for comparing floats? We could do a reachedTarget method which checks if delta between two floats is lower than an epsilon value? That's a more traditional way of comparing floats.

File: client/core/src/com/cows/game/controllers/TowerController.kt [hunk 3]
```diff
@@ -2,25 +2,29 @@ package com.cows.game.controllers
...
-
         model.rotation = MathUtils.atan2(directionToTarget.y, directionToTarget.x) * MathUtils.radDeg
+
+        currentBullet = BulletController(BulletModel(model.tileCoordinate.toVector2()))
+
     }
 
     val renderableView = TowerView(model)
 
+
```
Comment (AndreasWintherMoen) [inline client/core/src/com/cows/game/controllers/TowerController.kt:27]: Unnecessary empty line :( :(

**Review by AndreasWintherMoen — COMMENTED** [review]
Nice! Couple of minor suggestions (see comments)

File: client/core/src/com/cows/game/controllers/BulletController.kt (hunk 2 above)
Comment (fredrikbw) [reply client/core/src/com/cows/game/controllers/BulletController.kt:40]: @AndreasWintherMoen Yes that basically it, I just want to check if the new positions x or y is the 'same' as the previous one, I'll try with a reachedTarget method 👍

Comment (fredrikbw) [reply client/core/src/com/cows/game/controllers/BulletController.kt:40]: Resolving due to duplicate?

**Comment by AndreasWintherMoen** [comment]
I tried running it and it crashed? Apparently the bullet sprite isn't uploaded.

**Comment by fredrikbw** [comment]
Ok i don't know why that is, commited it again now

File: client/core/src/com/cows/game/controllers/BulletController.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+    override fun update(deltaTime: Float) {
+        if (model.hasHit) return
+        if (currentTarget == null) return
+        val targetPos = currentTarget!!.model.position
+        val oldX = model.position.x
+        val oldY = model.position.y
+
+        model.position.x += ((targetPos.x + 32f - model.position.x)) / model.movementSpeed
```
Comment (blauks) [inline client/core/src/com/cows/game/controllers/BulletController.kt:20]: Is this the bullet speed?

**Comment by blauks** [comment]
> I tried running it and it crashed? Apparently the bullet sprite isn't uploaded.

Remember to use the `24-tower-attack-action` branch and not `24-implement-tower-attack-action`:)

File: client/core/src/com/cows/game/controllers/BulletController.kt (hunk 4 above)
Comment (fredrikbw) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: It moves the bullet towards the target, and how much it moves depends on the bullet speed.

Comment (blauks) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: Ah okay, but 32f is the speed?

Comment (fredrikbw) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: no, 32f is to move it to the center of of the unit. Should probably have a getSize() or getCenterPosition() method in the Unit MVC if units will have different sizes. If this 32f is not added then the bullet would target the lower left corner of a unit

Comment (blauks) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: Okay doke

**Review by blauks — APPROVED** [review]
LGTM TO ME!

File: client/core/src/com/cows/game/controllers/BulletController.kt (hunk 4 above)
Comment (fredrikbw) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: image

Basically like this :)

Comment (blauks) [reply client/core/src/com/cows/game/controllers/BulletController.kt:20]: Nice!

**Review by Nicolaad — APPROVED** [review]
I think this looks good - well done! :D

**Pull request merged**